[FLINK-7381] [web] Decouple WebRuntimeMonitor from ActorGateway

tillrohrmann opened the pull request:
## What is the purpose of the change

This PR decouples the `WebRuntimeMonitor` from the `ActorGateway` by introducing the `JobManagerGateway` interface which can have multiple implementations. This is a preliminary step for the integration of the existing `WebRuntimeMonitor` with the Flip-6 `JobMaster`.

This PR is based #4486.

## Brief change log

- Extending the `JobManagerGateway` with methods to cover the requirements of the `WebRuntimeMonitor`
- Change `JobManagerRetriever` to return `JobManagerGateway` instead of `Tuple2<ActorGateway, Integer>`
- Adapt handlers to use the `JobManagerRetriever`
- Introduce `MetricQueryServiceRetriever` to retrieve `MetricQueryService` implementations running alongside with the JM and TM
- Introduce `MetricQueryServiceGateway` to abstract the implementation details of the `MetricQueryService` (e.g. Akka based)
- Pass `MetricQueryServiceRetriever` to `WebRuntimeMonitor` and `MetricFetcher`
- Adapt test classes to work with newly introduced interfaces
- Add `web.timeout` configuration option to control the WebRuntimeMonitor timeouts (see `WebMonitorOptions#WEB_TIMEOUT`

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (nop)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerGateway.java:
```diff
@@ -63,4 +70,103 @@
+	CompletableFuture<Optional<JobManagerMessages.ClassloadingProps>> requestClassloadingProps(JobID jobId, Time timeout);
+
+	/**
+	 * Requests the TaksManager instance registered under the given instanceId from the JobManager.
```
typo TaskManager

tillrohrmann replied:
Thanks for catching.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorUtils.java:
```diff
@@ -247,6 +262,8 @@ public static Path validateAndNormalizeUri(URI archiveDirUri) {
 	}
 
+
+
```
remove empty lines

tillrohrmann replied:
Will do.

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/WebMonitorOptions.java:
```diff
@@ -0,0 +1,38 @@
+	 * Timeout for asynchronous operations by the WebRuntimeMonitor
+	 */
+	public static final ConfigOption<Long> WEB_TIMEOUT = ConfigOptions
+		.key("web.timeout")
```
this config key is inconsistent with the existing web monitor options,

tillrohrmann replied:
Yes, see my comments above. I think the existing web monitor options are a bit misleading because the web monitor is not necessarily bound to the `JobManager`.

zentol commented:
This PR also needs a rebase.

tillrohrmann commented on docs/setup/config.md:
```diff
@@ -389,6 +389,10 @@ These parameters allow for advanced tuning. The default values are sufficient wh
 
 - `jobmanager.web.access-control-allow-origin`: Enable custom access control parameter for allow origin header, default is `*`.
 
+### Web Frontend
```
True but this is the `JobManager Web Frontend`. In the future (with Flip-6) there will be more than one web frontend component which is not necessarily tied to the `JobManager`. Therefore, I though about introducing some general web frontend configuration options. What do you think?

zentol replied:
I think that having some options tied to the jobmanager while others aren't will just cause more work down the line. Either add it the the jobmanager options now and move it in a follow-up, or separate the options now and rebase the PR on top of it.

The "let's refactor the rest later" mentality just doesn't work for us, see the translation to ConfigOptions which _still isn't done_.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/configuration/WebMonitorOptions.java:
```diff
@@ -0,0 +1,38 @@
+ * Configuration options for the WebRuntimeMonitor
+ */
+@PublicEvolving
+public class WebMonitorOptions {
```
True, but the timeout option is not job manager specific. In the future there will be multiple web frontend components running (one running next to the dispatcher, another one running next to a JobManager). Therefore, I though about introducing general options for the web monitor.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricFetcher.java:
```diff
@@ -101,127 +98,112 @@ public void update() {
+	 * @param queryServicePath specifying the QueryServiceGateway
+	 */
+	private void retrieveAndQueryMetrics(String queryServicePath) {
+		final CompletableFuture<MetricQueryServiceGateway> jmQueryServiceGatewayFuture = queryServiceRetriever.retrieveService(queryServicePath);
```
True, an artifact of a refactoring. Will change it.

tillrohrmann commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricFetcher.java:
```diff
@@ -101,127 +98,112 @@ public void update() {
+						LOG.debug("Fetching metrics failed.", t);
+					} else {
+						List<MetricDump> dumpedMetrics = deserializer.deserialize(result);
+						for (MetricDump metric : dumpedMetrics) {
```
Is this really how FLINK-7368 will be solved? I thought there is still some discussion ongoing.

zentol replied:
Yes, there's still a discussion about it, but that's not an argument against fixing it now to make it behave the way it was originally intended to.

tillrohrmann commented:
Thanks for the review @zentol. I've addressed most of your comments modulo the web frontend configuration options and rebased the PR onto the latest master.

I've tested the web frontend locally with a standalone cluster but not with Yarn and HA. Will try these configuration out as well.

tillrohrmann commented:
I did the additional tests running Flink on Yarn with and without HA and the web frontend worked.

zentol commented:
Thanks for testing things on yarn/HA.

As for the web monitor options, I do see the point of not coupling new options to the jobmanager. However, I think it's a bad idea to have some be tied to the jobmanager while others aren't. This smells like a repeat of the ConfigOption introduction, where we approached things with a "let's refactor the rest later" mentality, and that still isn't complete. I suggest to port all webmonitor options first (with deprecates keys obviously), and then rebase the PR on top of that.

tillrohrmann commented:
Alright, I'll port the web monitor options and rebase this PR onto that.

tillrohrmann commented:
@zentol I've created the [WebOptions PR](#4512) and rebased onto that as you've requested.

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/WebOptions.java:
```diff
@@ -149,6 +149,13 @@
+	/**
+	 * Timeout for asynchronous operations by the WebRuntimeMonitor
+	 */
+	public static final ConfigOption<Long> TIMEOUT = ConfigOptions
```
We should document the time-unit in some way.

tillrohrmann replied:
True. Will add it.

tillrohrmann commented:
Thanks for the review @zentol. Rebasing onto the latest master and if Travis gives green light, then I'll merge this PR.
